MapleServer2/Servers/Game/DungeonManager.cs

```diff
@@ -0,0 +1,132 @@
+        //        }
+        //        return mapInstanceId;
+
+        public DungeonSession CreateDungeonSession(int dungeonId, DungeonType dungeonType)
+        {
+            int dungeonSessionId = GetUniqueSessionId();
+            int dungeonInstanceId = GetMapInstanceId();
+            DungeonSession dungeonSession = new DungeonSession(dungeonSessionId, dungeonId, dungeonInstanceId, dungeonType);
+            DungeonSessionList.Add(dungeonSessionId, dungeonSession);
+            return dungeonSession;
+        }
+
+        public void RemoveDungeonSession(int dungeonSessionId)
+        {
+            if (!DungeonSessionList.ContainsKey(dungeonSessionId))
+            {
+                return;
+            }
+            int currentDungeonSessionId = DungeonSessionList[dungeonSessionId].SessionId;
+            int currentDungeonInstanceId = DungeonSessionList[dungeonSessionId].DungeonInstanceId;
+            RecyclableSessionIds.Add(currentDungeonSessionId);
+            RecyclableMapInstanceIds.Add(currentDungeonInstanceId);
+            DungeonSessionList.Remove(dungeonSessionId);
+        }
+
+        public int GetUniqueSessionId()
+        {
+            if (RecyclableSessionIds.Count > 0)
+            {
+                int sessionId = RecyclableSessionIds.First();
+                RecyclableSessionIds.Remove(sessionId);
+                return sessionId;
+            }
+            return SessionId++;
+        }
+
+        public DungeonSession GetDungeonSession(int dungeonSessionId)
+        {
+            //set dungeon of session id to x on swich dungeon
+            //if dungeonsession exists but id != dungeonid call dungeonid change
+            //update dungeonid or handle it somewhere else
+            if (!DungeonSessionList.ContainsKey(dungeonSessionId))
+            {
+                return null;
+            }
+            return DungeonSessionList[dungeonSessionId];
+        }
+
+        public DungeonSession GetDungeonSessionByInstance(int instanceId)
+        {
+            //TODO: if no dungeonsession is found, return null
+            return DungeonSessionList.FirstOrDefault(session => session.Value.DungeonInstanceId == instanceId).Value;
+        }
+
+        public bool DungeonUsesMap(DungeonSession dungeonSession, FieldManager fieldManager, Player player)
+        {
+            //fieldmanager is the map the player is coming from/that is to be released
+            if (dungeonSession != null) //is not null after entering dungeon via directory
+            {
+                //check map that is left: fieldmanager.mapId is the to be released map
+                if (dungeonSession.ContainsMap(fieldManager.MapId))  //map is a dungeon map
+                {
+                    //check map the player will travel to: lobby to dungeon and dungeon to lobby
+                    if (dungeonSession.ContainsMap(player.MapId))
+                    {
+                        return true;
```

Comment:
If you return true here, it'll return false in `MapleServer2/Tools/FieldManagerFactory.cs` **Release()** so the code will never reach `return managerList.Remove(fieldManager);`

Reply:
that is the point of it!
It is meant to not release an instance when a player travels from dungeon to lobby or lobby to dungeon! unless he leaves a dungeon map(lobby or dungeon map)

Reply:
but what if dungeon map doesn't have players anymore? it won't get released right? 🤔

Reply:
It will!
The else //travel destination is not a dungeon map
will handle this case. 
It will remove the dungeonsession, because being the last player in a dungeon and leaving from a dungeon map (lobby or dungeon) to a non dungeon map means the instance will be deleted (in official ms2)
Then, it will return false, saying the dungeon does not use the map, which runs the normal managers.Remove(...) and deletes the instance